File: OTEAServer/Controllers/EvidencesController.cs
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using OTEAServer.Misc;
using OTEAServer.Models;

namespace OTEAServer.Controllers
{
    /// <summary>
    /// Controller class for evidences operations
    /// Author: Pablo Ahíta del Barrio
    /// Version: 1
    /// </summary>
    
    
    [ApiController]
    [Route("Evidences")]
    public class EvidencesController : ControllerBase
    {
        /// <summary>
        /// Database context
        /// </summary>
        private readonly DatabaseContext _context;

        /// <summary>
        /// Class constructor
        /// </summary>
        /// <param name="context">Database context</param>
        public EvidencesController(DatabaseContext context)
        {
            _context = context;
        }


        /// <summary>
        /// Method that obtains all the evidences
        /// </summary>
        /// <param name="evaluationType">Evaluation type</param>
        /// <returns>Evidences list</returns>
        [HttpGet("all")]
        public IActionResult GetAll([FromQuery] string evaluationType)
        {
            try
            {
                var evidences = _context.Evidences
                    .Where(e=>e.evaluationType==evaluationType)
                    .OrderBy(e=>e.idIndicator)
                    .ThenBy(e=>e.idEvidence)
                    .ToList();
                return Ok(evidences);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that obtains from the database all the evidences of an indicator
        /// </summary>
        /// <param name="idIndicator">Indicator identifier</param>
        /// <param name="indicatorType">Indicator type</param>
        /// <param name="idSubSubAmbit">Second level division of the ambit</param>
        /// <param name="idSubAmbit">First level division of the ambit</param>
        /// <param name="idAmbit">Ambit identifier</param>
        /// <param name="indicatorVersion">Indicator version</param>
        /// <param name="evaluationType">Evaluation type</param>
        /// <returns>Evidences list</returns>
        [HttpGet("ind")]
        public IActionResult GetAllByIndicator([FromQuery] int idIndicator, [FromQuery] string indicatorType, [FromQuery] int idSubSubAmbit, [FromQuery] int idSubAmbit, [FromQuery] int idAmbit, [FromQuery] int indicatorVersion, [FromQuery] string evaluationType)
        {
            try
            {
                var evidences = _context.Evidences.Where(e => e.idIndicator == idIndicator && e.indicatorType == indicatorType && e.idSubSubAmbit == idSubSubAmbit && e.idSubAmbit == idSubAmbit && e.idAmbit == idAmbit && e.indicatorVersion == indicatorVersion && e.evaluationType==evaluationType).ToList();
                return Ok(evidences);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that obtains from the database an evidence
        /// </summary>
        /// <param name="idEvidence">Evidence identifier</param>
        /// <param name="idIndicator">Indicator identifier</param>
        /// <param name="indicatorType">Indicator type</param>
        /// <param name="idSubSubAmbit">Second level division of the ambit</param>
        /// <param name="idSubAmbit">First level division of the ambit</param>
        /// <param name="idAmbit">Ambit identifier</param>
        /// <param name="indicatorVersion">Indicator version</param>
        /// <param name="evaluationType">Evaluation type</param>
        /// <returns>Evidence if success, null if not</returns>
        [HttpGet("get")]
        public ActionResult<Evidence> Get([FromQuery] int idEvidence, [FromQuery] int idIndicator, [FromQuery] string indicatorType, [FromQuery] int idSubSubAmbit, [FromQuery] int idSubAmbit, [FromQuery] int idAmbit, [FromQuery] int indicatorVersion, [FromQuery] string evaluationType)
        {
            try
            {
                var evidence = _context.Evidences.FirstOrDefault(e => e.idEvidence == idEvidence && e.idIndicator == idIndicator && e.indicatorType == indicatorType && e.idSubSubAmbit == idSubSubAmbit && e.idSubAmbit == idSubAmbit && e.idAmbit == idAmbit && e.indicatorVersion == indicatorVersion && e.evaluationType==evaluationType);

                if (evidence == null)
                    return NotFound();

                return evidence;
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }



        /// <summary>
        /// Method that appends a new evidence to the database
        /// </summary>
        /// <param name="evidence">Evidence</param>
        /// <returns>Evidence if sucess, null if not</returns>
        [HttpPost]
        public IActionResult Create([FromBody] Evidence evidence)
        {
            try
            {
                _context.Evidences.Add(evidence);
                _context.SaveChanges();
                return CreatedAtAction(nameof(Get), new { id = evidence.idEvidence, idIndicator = evidence.idIndicator, type = evidence.indicatorType, idSubSubAmbit = evidence.idSubSubAmbit, idSubAmbit = evidence.idSubAmbit, idAmbit = evidence.idAmbit, version = evidence.indicatorVersion , evaluationType = evidence.evaluationType}, evidence);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that updates an existant evidence
        /// </summary>
        /// <param name="idEvidence">Evidence identifier</param>
        /// <param name="idIndicator">Indicator identifier</param>
        /// <param name="indicatorType">Indicator type</param>
        /// <param name="idSubSubAmbit">Second level division of the ambit</param>
        /// <param name="idSubAmbit">First level division of the ambit</param>
        /// <param name="idAmbit">Ambit identifier</param>
        /// <param name="indicatorVersion">Indicator version</param>
        /// <param name="evaluationType">Evaluation type</param>
        /// <param name="evidence">Evidence</param>
        /// <returns>Updated evidence if success, null if not</returns>
        [HttpPut]
        public IActionResult Update([FromQuery] int idEvidence, [FromQuery] int idIndicator, [FromQuery] string indicatorType, [FromQuery] int idSubSubAmbit, [FromQuery] int idSubAmbit, [FromQuery] int idAmbit, [FromQuery] int indicatorVersion, [FromQuery] string evaluationType, [FromBody] Evidence evidence)
        {
            try
            {
                if (idEvidence != evidence.idEvidence || idIndicator != evidence.idIndicator || idSubSubAmbit != evidence.idSubSubAmbit || idSubAmbit != evidence.idSubAmbit || idAmbit != evidence.idAmbit || indicatorType != evidence.indicatorType || evaluationType!=evidence.evaluationType)
                    return BadRequest();

                var existingEvidence = _context.Evidences.FirstOrDefault(e => e.idEvidence == idEvidence && e.idIndicator == idIndicator && e.indicatorType == indicatorType && e.idSubSubAmbit == idSubSubAmbit && e.idSubAmbit == idSubAmbit && e.idAmbit == idAmbit && e.indicatorVersion == indicatorVersion && e.evaluationType==evaluationType);
                if (existingEvidence is null)
                    return NotFound();

                existingEvidence.descriptionEnglish = evidence.descriptionEnglish;
                existingEvidence.descriptionSpanish = evidence.descriptionSpanish;
                existingEvidence.descriptionFrench = evidence.descriptionFrench;
                existingEvidence.descriptionBasque = evidence.descriptionBasque;
                existingEvidence.descriptionCatalan = evidence.descriptionCatalan;
                existingEvidence.descriptionDutch = evidence.descriptionDutch;
                existingEvidence.descriptionGalician = evidence.descriptionGalician;
                existingEvidence.descriptionGerman = evidence.descriptionGerman;
                existingEvidence.descriptionItalian = evidence.descriptionItalian;
                existingEvidence.descriptionPortuguese = evidence.descriptionPortuguese;
                existingEvidence.evidenceValue = evidence.evidenceValue;
                existingEvidence.evaluationType = evaluationType;

                _context.SaveChanges();

                return Ok(existingEvidence);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }

        }

        /// <summary>
        /// Method that deletes an evidence
        /// </summary>
        /// <param name="idEvidence">Evidence identifier</param>
        /// <param name="idIndicator">Indicator identifier</param>
        /// <param name="indicatorType">Indicator type</param>
        /// <param name="idSubSubAmbit">Second level division of the ambit</param>
        /// <param name="idSubAmbit">First level division of the ambit</param>
        /// <param name="idAmbit">Ambit identifier</param>
        /// <param name="indicatorVersion">Indicator version</param>
        /// <returns>Deleted evidence if success, null if not</returns>
        [HttpDelete]
        public IActionResult Delete([FromQuery] int idEvidence, [FromQuery] int idIndicator, [FromQuery] string indicatorType, [FromQuery] int idSubSubAmbit, [FromQuery] int idSubAmbit, [FromQuery] int idAmbit, [FromQuery] int indicatorVersion, [FromQuery] string evaluationType)
        {
            try
            {
                var evidence = _context.Evidences.FirstOrDefault(e => e.idEvidence == idEvidence && e.idIndicator == idIndicator && e.indicatorType == indicatorType && e.idSubSubAmbit == idSubSubAmbit && e.idSubAmbit == idSubAmbit && e.idAmbit == idAmbit && e.indicatorVersion == indicatorVersion && e.evaluationType == evaluationType);

                if (evidence is null)
                    return NotFound();

                _context.Evidences.Remove(evidence);
                _context.SaveChanges();

                return NoContent();
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }

        }
    }
}

File: connection/src/main/java/otea/connection/controller/EvidencesController.java
package otea.connection.controller;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import cli.indicators.Evidence;
import otea.connection.ConnectionClient;
import otea.connection.api.EvidencesApi;
import retrofit2.Call;
import retrofit2.Response;
import retrofit2.http.GET;
import retrofit2.http.Query;

/**
 * Controller class for evidences operations
 *
 * @author Pablo Ahíta del Barrio
 * @version 1
 * */
public class EvidencesController {

    /**Evidences api to connect to the server*/
    private static EvidencesApi api;

    /**Evidence controller instance*/
    private static EvidencesController instance;

    /**Class controller*/
    private EvidencesController(){
        api= ConnectionClient.getInstance().getRetrofit().create(EvidencesApi.class);
    }

    /**
     * Method that obtains the singleton instance of the controller
     *
     * @return Controller instance
     * */
    public static EvidencesController getInstance(){
        if(instance==null){
            synchronized (EvidencesController.class){
                if(instance==null){
                    instance=new EvidencesController();
                }
            }
        }
        return instance;
    }

    /**Refresh API*/
    public static void refreshApi(){
        instance=new EvidencesController();
    }

    /**
     * Method that obtains all evidences from the database
     *
     * @return Evidences list
     * */
    public static List<Evidence> GetAll(){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<List<Evidence>> callable = new Callable<List<Evidence>>() {
            @Override
            public List<Evidence> call() throws Exception {
                Call<List<Evidence>> call = api.GetAll();
                Response<List<Evidence>> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<List<Evidence>> future = executor.submit(callable);
            List<Evidence> list = future.get();
            executor.shutdown();
            return list;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }


    /**
     * Method that obtains from the database all the evidences of an indicator
     *
     * @param idIndicator - Indicator identifier
     * @param indicatorType - Indicator type
     * @param idSubSubAmbit - Second level division of the ambit
     * @param idSubAmbit  - First level division of the ambit
     * @param idAmbit - Ambit identifier
     * @param indicatorVersion - Indicator version
     * @return Evidences list
     * */
    public static List<Evidence> GetAllByIndicator(int idIndicator, String indicatorType, int idSubSubAmbit, int idSubAmbit, int idAmbit, int indicatorVersion){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<List<Evidence>> callable = new Callable<List<Evidence>>() {
            @Override
            public List<Evidence> call() throws Exception {
                Call<List<Evidence>> call = api.GetAllByIndicator(idIndicator, indicatorType, idSubSubAmbit, idSubAmbit, idAmbit, indicatorVersion);
                Response<List<Evidence>> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<List<Evidence>> future = executor.submit(callable);
            List<Evidence> list = future.get();
            executor.shutdown();
            return list;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that obtains from the database an evidence
     *
     * @param idEvidence - Evidence identifier
     * @param idIndicator - Indicator identifier
     * @param indicatorType - Indicator type
     * @param idSubSubAmbit - Second level division of the ambit
     * @param idSubAmbit  - First level division of the ambit
     * @param idAmbit - Ambit identifier
     * @param indicatorVersion - Indicator version
     * @return Evidence if success, null if not
     * */
    public static Evidence Get(int idEvidence, int idIndicator, String indicatorType, int idSubSubAmbit, int idSubAmbit, int idAmbit, int indicatorVersion){
    ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<Evidence> callable = new Callable<Evidence>() {
            @Override
            public Evidence call() throws Exception {
                Call<Evidence> call = api.Get(idEvidence,idIndicator,indicatorType,idSubSubAmbit,idSubAmbit,idAmbit,indicatorVersion);
                Response<Evidence> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<Evidence> future = executor.submit(callable);
            Evidence result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that appends a new evidence to the database
     *
     * @param evidence - Evidence
     * @return Evidence if sucess, null if not
     * */
    public static Evidence Create(Evidence evidence){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<Evidence> callable = new Callable<Evidence>() {
            @Override
            public Evidence call() throws Exception {
                Call<Evidence> call = api.Create(evidence);
                Response<Evidence> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<Evidence> future = executor.submit(callable);
            Evidence result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that updates an existant evidence
     *
     * @param idEvidence - Evidence identifier
     * @param idIndicator - Indicator identifier
     * @param indicatorType - Indicator type
     * @param idSubSubAmbit - Second level division of the ambit
     * @param idSubAmbit  - First level division of the ambit
     * @param idAmbit - Ambit identifier
     * @param indicatorVersion - Indicator version
     * @param evidence - Evidence
     * @return Updated evidence if success, null if not
     * */
    public static Evidence Update(int idEvidence, int idIndicator, String indicatorType, int idSubSubAmbit, int idSubAmbit, int idAmbit, int indicatorVersion, Evidence evidence) {
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<Evidence> callable = new Callable<Evidence>() {
            @Override
            public Evidence call() throws Exception {
                Call<Evidence> call = api.Update(idEvidence,idIndicator,indicatorType,idSubSubAmbit,idSubAmbit,idAmbit,indicatorVersion,evidence);
                Response<Evidence> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<Evidence> future = executor.submit(callable);
            Evidence result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that deletes an evidence
     *
     * @param idEvidence - Evidence identifier
     * @param idIndicator - Indicator identifier
     * @param indicatorType - Indicator type
     * @param idSubSubAmbit - Second level division of the ambit
     * @param idSubAmbit  - First level division of the ambit
     * @param idAmbit - Ambit identifier
     * @param indicatorVersion - Indicator version
     * @return Deleted evidence if success, null if not
     * */
    public static Evidence Delete(int idEvidence, int idIndicator, String indicatorType, int idSubSubAmbit, int idSubAmbit, int idAmbit, int indicatorVersion) {
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<Evidence> callable = new Callable<Evidence>() {
            @Override
            public Evidence call() throws Exception {
                Call<Evidence> call = api.Delete(idEvidence,idIndicator,indicatorType,idSubSubAmbit,idSubAmbit,idAmbit,indicatorVersion);
                Response<Evidence> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<Evidence> future = executor.submit(callable);
            Evidence result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

}
